pennylane/gradients/gradient_transform.py

```diff
@@ -394,80 +397,92 @@ def reorder_grads(grads, tape_specs):
-        def _reshape(x):
-            return qml.math.reshape(x, (1,) if x.shape == () else (1, -1))
-
-        if not (multi_meas or has_partitioned_shots):
-            # Single parameter, single measurements, no shot vector
-            return tdot(_reshape(qjac), cjac)
-
-        if not (multi_meas and has_partitioned_shots):
-            # Single parameter, multiple measurements or shot vector, but not both
-            return tuple(tdot(_reshape(q), cjac) for q in qjac)
-
-        # Single parameter, multiple measurements, and shot vector
-        return tuple(tuple(tdot(_reshape(_q), cjac) for _q in q) for q in qjac)
-
-    if not multi_meas:
-        # Multiple parameters, single measurement
-        qjac = stack(qjac)
-        if not cjac_is_tuple:
-            cjac = stack(cjac)
-            if has_partitioned_shots:
-                return tuple(tdot(stack(q), cjac) for q in qjac)
-            return tdot(qjac, cjac)
-        if has_partitioned_shots:
-            return tuple(tuple(tdot(q, c) for c in cjac if c is not None) for q in qjac)
-        return tuple(tdot(qjac, c) for c in cjac if c is not None)
-
-    # Multiple parameters, multiple measurements
-    if not cjac_is_tuple:
-        cjac = stack(cjac)
-        if has_partitioned_shots:
-            return tuple(tuple(tdot(stack(_q), cjac) for _q in q) for q in qjac)
-        return tuple(tdot(stack(q), cjac) for q in qjac)
-    if has_partitioned_shots:
-        return tuple(
-            tuple(tuple(tdot(stack(_q), c) for c in cjac if c is not None) for _q in q)
-            for q in qjac
-        )
-    return tuple(tuple(tdot(stack(q), c) for c in cjac if c is not None) for q in qjac)
+    if tape.shots.has_partitioned_shots:
+        return tuple(_single_shots_contraction(qjac_s, cjac, tape) for qjac_s in qjac)
+
+    return _single_shots_contraction(qjac, cjac, tape)
+
+
+def _single_shots_contraction(qjac, cjac, tape):
+    single_trainable_param = len(tape.trainable_params) == 1
+    if len(tape.measurements) == 1:
+        return _single_meas_contraction(qjac, cjac, single_trainable_param)
+
+    return tuple(_single_meas_contraction(qj, cjac, single_trainable_param) for qj in qjac)
+
+
+def _single_meas_contraction(qjac, cjac, single_trainable_param):
+    if single_trainable_param:
+        # return spec squeezed out singleton dimension, so we need to add it back in
+        qjac = qml.math.expand_dims(qjac, axis=0)
+    qjac = qml.math.stack(qjac)
+
+    # pytrees are easiest way to handle different argnum conventions: argnum=None, argnum=0, argnum=[0], etc.
+    cjac_leaves, struct = flatten(cjac)
+
+    contracted = []
+    for cjac_for_arg in cjac_leaves:
+        tdot = qml.math.tensordot(cjac_for_arg, qjac, axes=[[0], [0]])
+        # not sure how to rationalize this tranposition, but its needed
+        contracted.append(tdot if len(tdot.shape) == 0 else qml.math.moveaxis(tdot, -1, 0))
+
+    return unflatten(contracted, struct)

```

Comment:
```suggestion
    def _wrapped_contraction(x):
        """Tensordot with qjac, and then cyclic transpose the last axix to 'wrap' the result"""
        tdot = qml.math.tensordot(x, qjac, axes=[[0], [0]])
        # not sure how to rationalize this tranposition, but its needed
        return tdot if len(tdot.shape) <= 1 else qml.math.moveaxis(tdot, -1, 0)

    contracted = [for cjac_for_arg in cjac_leaves]

    return unflatten(contracted, struct)
```
Idk maybe list generation can be bit faster if tensors endup large :smiley: feel free to ignore

Reply:
Got rid of the `moveaxis` because it turns out I was just contracting in the wrong order.